zebra-network/src/peer_set/initialize.rs

```diff
@@ -359,26 +361,38 @@ where
 /// `peerset_initial_target_size`.
 ///
 /// The result is randomly chosen entries from the provided set of addresses.
-async fn limit_initial_peers(config: &Config) -> HashSet<SocketAddr> {
-    let initial_peers = config.initial_peers().await;
-    let initial_peer_count = initial_peers.len();
+async fn limit_initial_peers(
+    config: &Config,
+    mut address_book_updater: mpsc::Sender<MetaAddrChange>,
+) -> HashSet<SocketAddr> {
+    let all_peers = config.initial_peers().await;
+    let peers_count = all_peers.len();
+
+    if peers_count <= config.peerset_initial_target_size {
+        return all_peers;
+    }
 
     // Limit the number of initial peers to `config.peerset_initial_target_size`
-    if initial_peer_count > config.peerset_initial_target_size {
-        info!(
-            "Limiting the initial peers list from {} to {}",
-            initial_peer_count, config.peerset_initial_target_size
-        );
-    }
+    info!(
+        "Limiting the initial peers list from {} to {}",
+        peers_count, config.peerset_initial_target_size
+    );
 
-    let initial_peers_vect: Vec<SocketAddr> = initial_peers.iter().copied().collect();
+    // Split all the peers into the `initial_peers` that will be returned and
+    // `unused_peers` that will be sent to the address book.
+    let mut all_peers: Vec<SocketAddr> = all_peers.iter().copied().collect();
+    let (initial_peers, unused_peers) =
+        all_peers.partial_shuffle(&mut rand::thread_rng(), config.peerset_initial_target_size);
+
+    // Send the unused peers to the address book.
+    for peer in unused_peers {
+        let peer_addr = MetaAddr::new_initial_peer(*peer);
+        // `send` only waits when the channel is full.
+        // The address book updater is a separate task, so we will only wait for a short time.
+        let _ = address_book_updater.send(peer_addr).await;
+    }
 
-    // TODO: add unused peers to the AddressBook (#2931)
-    //       https://docs.rs/rand/0.8.4/rand/seq/trait.SliceRandom.html#tymethod.partial_shuffle
-    initial_peers_vect
-        .choose_multiple(&mut rand::thread_rng(), config.peerset_initial_target_size)
-        .copied()
-        .collect()
+    initial_peers.iter().copied().collect()
```

Comment:
Minor improvement: same here:

```suggestion
    initial_peers.into_iter().collect()
```

Reply:
Oh no! I didn't realize `partial_shuffle` returned slices. That means `.into_iter()` will borrow the items, because a slice is a borrowed memory section. I'm very sorry for this :(

Reply:
Hm, the compiler says: "value of type `std::collections::HashSet<std::net::SocketAddr>` cannot be built from `std::iter::Iterator<Item=&mut std::net::SocketAddr>`" after I used this.

Reply:
Oh, I commented without noticing your last comment.

Reply:
Although, I still don't feel like I've fully grasped the issue here. :/ Especially the error message.